enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,355 @@
+      action: Allow
+  - fromCIDRS:
+       - 10:10::1/64
+    protocol:
+    icmpv6:
+      type: 128 # Ping request
+      action: Deny
+```
+
+- eBPF data structures and types
+```c
+#define DENY XDP_DROP
+#define ALLOW XDP_PASS
+
+struct ruleInfo {
+    __u32 ruleId;
+    __u8 protocol;
+    __u32 srcAddr;
+    __u32 srcMask;
+    __u16 dstPort;
+    __u8 icmpType;
+    __u8 icmpCode;
+    __u8 action;
+} __packed;
+
+#define MAX_INSIDE_TARGETS   (1024)
+#define MAX_RULES_PER_TARGET (100)
+
+struct bpf_map_def SEC("maps") ingressRulesTbl = {
+    .type = BPF_MAP_TYPE_LPM_TRIE,
+    .key_size = sizeof(struct bpf_lpm_ip_trie),
+    .value_size = sizeof(struct rulesStore) + (sizeof(struct ruleInfo)*MAX_RULES_PER_TARGET),
+    .max_entries = MAX_INSIDE_TARGETS,
+    .map_flags = BPF_F_NO_PREALLOC,
+};
+
+// using Longest perfix match in case of overlapping CIDRs we need to match to the more specific CIDR.
+struct bpf_lpm_trie_ip_key {
+    __u32 prefixlen;
+    union {
+        __u8 data[4];
+        __u8 data[16];
+    } u;
+} __packed;
+
+struct rulesStore {
+    __u32 numRules;
+    struct ruleInfo rulesInfo[0];
+} __packed;
+```
+
+![eBPF datastructures](./ebpf-structures.png)
+
+### User Stories
+
+#### Story 1
+
+### API Extensions
+
+A new cluster scoped CRD is introduced to support this feature this CRD will be managed by `ClusterNetworkOperator`
+```golang
+type IngressNodeFirewallFromCIDR struct {
+	// FromCIDR is the IPv4/IPv6 CIDR from which we apply node firewall rule
+	// +kubebuilder:validation:Pattern=`^(([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])/([0-9]|[12][0-9]|3[0-2])$`
+	FromCIDR string `json:"fromCIDR" protobuf:"bytes,1,opt,name=fromCIDR"`
+}
+
+// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
+type IngressNodeFirewallProtocolRule struct {
+	// Port is the port number to which clients should send traffic to.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=65535
+	// +kubebuilder:validation:Minimum:=1
+	Port int32 `json:"port" protobuf:"varint,1,opt,name=port"`
+
+	// IcmpType define ICMP Type Numbers (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=43
+	// +kubebuilder:validation:Minimum:=0
+	IcmpType int8 `json:"icmpType" protobuf:"varint,2,opt,name=icmpType"`
+
+	// IcmpCode define ICMP Code ID (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=16
+	// +kubebuilder:validation:Minimum:=0
+	IcmpCode int8 `json:"icmpCode" protobuf:"varint,2,opt,name=icmpCode"`
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="ICMP";"ICMPV6";"TCP";"UDP";"SCTP"
+	Protocol IngressNodeFirewallRuleProtocolType `json:"protocol" protobuf:"bytes,2,opt,name=protocol,casttype=IngressNodeFirewallRuleProtocolType"`
+
+	// Action can be Allow or Deny.
```

Comment:
need to clarify what happens if one rule allows traffic and another denies it. (eg, the first rule takes effect, the last rule takes effect, etc)

Reply:
rules should be distinct we can't have
```
protocol:
- tcp:
  port: 80
  action: allow
- tcp:
  port: 80
  action: deny
```
validation webhook should reject this config

Reply:
exactly that is the case we will match the more specific rule otherwise hit the fallback rule if it explicit if not the implicit 0.0.0.0/0 that deny all protocols (ofcourse outside the failsafe rules) will take effect.